corehq/blobs/migrate.py

```diff
@@ -127,122 +123,142 @@
-    once may be very innefficient (it may copy the full content of every
-    blob each time), depending on the implementation of the databases
-    being migrated.
-    """
-    db = get_blob_db()
-    if not isinstance(db, MigratingBlobDB):
-        raise MigrationError(
-            "Expected to find migrating blob db backend (got %r)" % db)
+class BaseDocMigrator(object):
 
-    not_found = 0
-    start = datetime.now()
-    with open(filename, 'r') as f:
-        for n, line in enumerate(f):
-            if n % 100 == 0:
-                print_status(n + 1, total, datetime.now() - start)
-            doc = json.loads(line)
-            obj = doc_type_map[doc["doc_type"]].wrap(doc)
-            bucket = obj._blobdb_bucket()
-            assert obj.external_blobs == obj.blobs, (obj.external_blobs, obj.blobs)
-            for name, meta in obj.blobs.iteritems():
-                try:
-                    content = db.old_db.get(meta.id, bucket)
-                except NotFound:
-                    not_found += 1
-                else:
-                    with content:
-                        db.copy_blob(content, meta.info, bucket)
-    if not_found:
-        print("{} documents were not found in the old blob database. It is "
-              "possible that some blobs were deleted as part of normal "
-              "operation during the migration if the migration took a long "
-              "time. However, it may be cause for concern if a majority of "
-              "the total number of migrated blobs were not found."
-              .format(not_found))
-    return total - not_found, 0
+    #blobs_key = None  # Abstract: doc key to be tested for migration status.
+    # If this key contains a falsy value the document will not be migrated.
+
+    # If true, load attachment content before migrating.
+    load_attachments = False
+
+    def migrate(self, doc, couchdb):
+        """Migrate a single document
+
+        :param doc: The document dict to be migrated.
+        :param couchdb: Couchdb database in which to save migrated doc.
+        :returns: True if doc was migrated else False. If this returns False
+        the document migration will be retried later.
+        """
+        raise NotImplementedError("abstract method")
+
+    def after_migration(self):
+        pass
+
+
+class CouchAttachmentMigrator(BaseDocMigrator):
 
-migrate_blob_db_backend.blobs_key = "external_blobs"
+    blobs_key = "_attachments"
+    load_attachments = True
+
+    def migrate(self, doc, couchdb):
+        attachments = doc.pop("_attachments")
+        external_blobs = doc.setdefault("external_blobs", {})
+        obj = BlobHelper(doc, couchdb)
+        try:
+            with obj.atomic_blobs():
+                for name, data in list(attachments.iteritems()):
+                    if name in external_blobs:
+                        continue  # skip attachment already in blob db
+                    obj.put_attachment(name=name, **data)
```

Comment:
yes, but there should always be a change since there are always attachments to migrate when this function is called.